### PR TITLE
PolyVox improvements -- depends on #5008

### PR DESCRIPTION
**Pull request by sethalves**
edit.js has some support for polyvox entities.  lots of bounds and sanity checks.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: ea3f002a65e3ca2942cd0f965bf9b0ff4fd52245 available at: https://gist.github.com/4095725868581d134fc2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5015.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5015.dmg

File: examples/edit.js [hunk 1]
```diff
@@ -266,6 +276,7 @@ var toolBar = (function () {
         toolBar.showTool(newTextButton, doShow);
         toolBar.showTool(newWebButton, doShow);
         toolBar.showTool(newZoneButton, doShow);
+	toolBar.showTool(newPolyVoxButton, doShow);
     };
```
Comment (AndrewMeadows) [inline examples/edit.js]: Bad indentation at line 279.

File: examples/html/entityProperties.html [hunk 2]
```diff
@@ -354,6 +354,12 @@
             var elZoneAtmosphereScatteringWavelengthsZ = document.getElementById("property-zone-atmosphere-scattering-wavelengths-z");
             var elZoneAtmosphereHasStars = document.getElementById("property-zone-atmosphere-has-stars");
 
+	    var elPolyVoxSelections = document.querySelectorAll(".poly-vox-section");
+	    var elVoxelVolumeSizeX = document.getElementById("property-voxel-volume-size-x");
+            var elVoxelVolumeSizeY = document.getElementById("property-voxel-volume-size-y");
+            var elVoxelVolumeSizeZ = document.getElementById("property-voxel-volume-size-z");
+	    var elVoxelSurfaceStyle = document.getElementById("property-voxel-surface-style");
```
Comment (AndrewMeadows) [inline examples/html/entityProperties.html]: Hrm... more bad indentation.  It appears there are some **tabs** in this file.

File: examples/voxels.js [hunk 3]
```diff
@@ -35,6 +41,9 @@ function keyReleaseEvent(event) {
     if (event.text == "CONTROL") {
         controlHeld = false;
     }
+    if (event.text == "SHIFT") {
+        shiftHeld = false;
+    }
 }
```
Comment (AndrewMeadows) [inline examples/voxels.js]: Tabs in this file too.

File: libraries/entities-renderer/src/RenderablePolyVoxEntityItem.cpp [hunk 4]
```diff
@@ -33,22 +33,67 @@
...
+
+
+bool inBounds(const PolyVox::SimpleVolume<uint8_t>* vol, int x, int y, int z) {
+    // x, y, z are in polyvox volume coords
+    if (x < 0 || y < 0 || z < 0 ||
+        x >= vol->getWidth() || y >= vol->getHeight() || z >= vol->getDepth()) {
+        return false;
+    }
+    return true;
+}
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderablePolyVoxEntityItem.cpp]: This function could be collapsed into one line that would avoid a logic branch:

return !(x < 0 || y < 0 || z < 0 || x >= vol->getWidth() || y >= vol->getHeight() || z >= vol->getDepth());

File: libraries/entities-renderer/src/RenderablePolyVoxEntityItem.cpp [hunk 5]
```diff
@@ -73,57 +173,138 @@ void RenderablePolyVoxEntityItem::setVoxelData(QByteArray voxelData) {
...
 
-glm::mat4 RenderablePolyVoxEntityItem::voxelToWorldMatrix() const {
+glm::mat4 RenderablePolyVoxEntityItem::voxelToLocalMatrix() const {
     glm::vec3 scale = _dimensions / _voxelVolumeSize; // meters / voxel-units
-    glm::mat4 scaled = glm::scale(glm::mat4(), scale);
-    glm::mat4 centerToCorner = glm::translate(scaled, _voxelVolumeSize / -2.0f);
+    glm::vec3 center = getCenter();
+    glm::vec3 position = getPosition();
+    glm::vec3 positionToCenter = center - position;
+    positionToCenter -= _dimensions * glm::vec3(0.5f,0.5f,0.5f) - getSurfacePositionAdjustment();
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderablePolyVoxEntityItem.cpp]: Need spaces between floats in glm::vec3 ctor.

BTW, glm::vec3 has a ctor that takes a single float and copies it to all components, so glm::vec3(0.5f) produces the same result as glm::vec3(0.5f, 0.5f, 0.5f).

File: libraries/entities-renderer/src/RenderablePolyVoxEntityItem.cpp [hunk 6]
```diff
@@ -73,57 +173,138 @@ void RenderablePolyVoxEntityItem::setVoxelData(QByteArray voxelData) {
...
+    switch (_voxelSurfaceStyle) {
+        case PolyVoxEntityItem::SURFACE_MARCHING_CUBES:
+            return scale / 2.0f;
+        case PolyVoxEntityItem::SURFACE_EDGED_CUBIC:
+            return scale / -2.0f;
+        case PolyVoxEntityItem::SURFACE_CUBIC:
+            return scale / 2.0f;
+    }
+    return glm::vec3(0, 0, 0);
+}
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderablePolyVoxEntityItem.cpp]: There are implicit casts happening in the glm::vec3 ctor in line 186.  0 --> 0.0f

File: libraries/entities/src/PolyVoxEntityItem.cpp [hunk 7]
```diff
@@ -41,6 +42,40 @@ PolyVoxEntityItem::PolyVoxEntityItem(const EntityItemID& entityItemID, const Ent
...
+
+    if (_voxelVolumeSize.z < 1) {
+        qDebug() << "PolyVoxEntityItem::setVoxelVolumeSize clamping z of" << _voxelVolumeSize.z << "to 1";
+        _voxelVolumeSize.z = 1;
+    }
+    if (_voxelVolumeSize.z > MAX_VOXEL_DIMENSION) {
+        qDebug() << "PolyVoxEntityItem::setVoxelVolumeSize clamping z of" << _voxelVolumeSize.z << "to max";
+        _voxelVolumeSize.z = MAX_VOXEL_DIMENSION;
+    }
+}
```
Comment (AndrewMeadows) [inline libraries/entities/src/PolyVoxEntityItem.cpp]: If you're willing to forego the debug output you could achieve the clamping in just one line:

_voxelVolumeSize = glm::clamp(voxelVolumeSize, 1, MAX_VOXEL_DIMENSION);

**Comment by hifi-gustavo** [comment]
Build console output for commit: d201f7791e29f032a28d40a24d875355bba309da available at: https://gist.github.com/4095725868581d134fc2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5015.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5015.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 11eb073bcfd3aebf404a454ad50596d882572aa4 available at: https://gist.github.com/4095725868581d134fc2<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5015.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-5015.dmg

**Pull request merged**